### PR TITLE
Hamiltonian Cycle

### PR DESCRIPTION
**Pull request by Mysterious786**
### **Describe your change:**

- [x] Add an algorithm?
- [ ] Fix a bug or typo in an existing algorithm?
- [x] Documentation change?

### **Checklist:**

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] I know that pull requests will not be merged if they fail the automated tests.
- [x] This PR only changes one algorithm file. To ease review, please open separate PRs for separate algorithms.
- [x] All new Java files are placed inside an existing directory.
- [x] All filenames are in all uppercase characters with no spaces or dashes.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.
- [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Review by Mysterious786 — COMMENTED** [review]
Hamiltonian Cycle

File: src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+{
+   
+    final int V = 5;
+    int path[];
+ 
+    /* A utility function to check if the vertex v can be
+       added at index 'pos'in the Hamiltonian Cycle
+       constructed so far (stored in 'path[]') */
+    boolean isSafe(int v, int graph[][], int path[], int pos)
+    {
```
Comment (debasishbsws) [inline src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java:14]:
```suggestion
    boolean isSafe(int v, int graph[][], int path[], int pos) {
```

File: src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            if (path[i] == v)
+                return false;
+ 
+        return true;
+    }
+ 
+    /* A recursive utility function to solve hamiltonian
+       cycle problem */
+    boolean hamCycleUtil(int graph[][], int path[], int pos)
+    {
```
Comment (debasishbsws) [inline src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java:33]:
```suggestion
    boolean hamCycleUtil(int graph[][], int path[], int pos) {
```

File: src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        /* base case: If all vertices are included in
+           Hamiltonian Cycle */
+        if (pos == V)
+        {
```
Comment (debasishbsws) [inline src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java:37]:
```suggestion
        if (pos == V) {
```

File: src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+                return true;
+            else
+                return false;
+        }
+ 
+        // Try different vertices as a next candidate in
+        // Hamiltonian Cycle. We don't try for 0 as we
+        // included 0 as starting point in hamCycle()
+        for (int v = 1; v < V; v++)
+        {
```
Comment (debasishbsws) [inline src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java:50]:
```suggestion
        for (int v = 1; v < V; v++) {
```

File: src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            /* Check if this vertex can be added to Hamiltonian
+               Cycle */
+            if (isSafe(v, graph, path, pos))
+            {
```
Comment (debasishbsws) [inline src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java:54]:
```suggestion
            if (isSafe(v, graph, path, pos)) {
```

File: src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    }
+ 
+    /* This function solves the Hamiltonian Cycle problem using
+       Backtracking. It mainly uses hamCycleUtil() to solve the
+       problem. It returns false if there is no Hamiltonian Cycle
+       possible, otherwise return true and prints the path.
+       Please note that there may be more than one solutions,
+       this function prints one of the feasible solutions. */
+    int hamCycle(int graph[][])
+    {
```
Comment (debasishbsws) [inline src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java:79]:
```suggestion
    int hamCycle(int graph[][]) {
```

File: src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        for (int i = 0; i < V; i++)
+            path[i] = -1;
+ 
+        /* Let us put vertex 0 as the first vertex in the path.
+           If there is a Hamiltonian Cycle, then the path can be
+           started from any point of the cycle as the graph is
+           undirected */
+        path[0] = 0;
+        if (hamCycleUtil(graph, path, 1) == false)
+        {
```
Comment (debasishbsws) [inline src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java:90]:
```suggestion
        if (hamCycleUtil(graph, path, 1) == false) {
```

File: src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+            return 0;
+        }
+ 
+        printSolution(path);
+        return 1;
+    }
+ 
+    /* A utility function to print solution */
+    void printSolution(int path[])
+    {
```
Comment (debasishbsws) [inline src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java:101]:
```suggestion
    void printSolution(int path[]) {
```

File: src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+            System.out.print(" " + path[i] + " ");
+ 
+        // Let us print the first vertex again to show the
+        // complete cycle
+        System.out.println(" " + path[0] + " ");
+    }
+ 
+    // driver program to test above function
+    public static void main(String args[])
+    {
```
Comment (debasishbsws) [inline src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java:114]:
```suggestion
    public static void main(String args[]) {
```

File: src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java [hunk 10]
```diff
@@ -0,0 +1,149 @@
+package com.thealgorithms.backtracking;
```
Comment (debasishbsws) [inline src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java:1]: Add comments about the algorithm, and put some links where people can read and know about the algorithm

**Review by debasishbsws — CHANGES_REQUESTED** [review]
Little formatting will look good

**Comment by Mysterious786** [comment]
Thanks for formatting suggestion

**Review by debasishbsws — APPROVED** [review]
Looks Good

File: src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java [hunk 11]
```diff
@@ -0,0 +1,140 @@
+package com.thealgorithms.backtracking;
+
```
Comment (Mysterious786) [inline src/main/java/com/thealgorithms/backtracking/HamiltonianCycle.java:2]: https://www.geeksforgeeks.org/hamiltonian-cycle-backtracking-6/

**Review by siriak — CHANGES_REQUESTED** [review]
The code looks good, could you please add some JUnit tests? (see https://github.com/TheAlgorithms/Java/tree/master/src/test/java/com/thealgorithms)

**Comment by raisahab9125** [comment]
lol its annoying     **think i  got stucked** 
@BurkHufnagel #3566 #3566 #3620 #3797

**Pull request closed without merging**